maze_transformer/training/config.py

```diff
@@ -210,7 +216,81 @@ def transformer_config(self) -> HookedTransformerConfig:
             d_vocab=len(self.dataset_cfg.token_arr),
         )
 
+    def transformer_config(self) -> HookedTransformerConfig:
+        warnings.warn(
+            "transformer_config is deprecated, use hooked_transformer_cfg instead",
+            DeprecationWarning,
+        )
+        return self.hooked_transformer_cfg
+
     def create_model(self) -> HookedTransformer:
         return HookedTransformer(
-            cfg=self.transformer_config(), tokenizer=self.tokenizer
+            cfg=self.hooked_transformer_cfg,
+            tokenizer=self.tokenizer,
+        )
+
+    def create_model_zanj(self) -> ZanjHookedTransformer:
+        return ZanjHookedTransformer(self)
+
+
+@set_config_class(ConfigHolder)
+class ZanjHookedTransformer(ConfiguredModel, HookedTransformer):
+    """A hooked transformer that is configured by a ConfigHolder
+
+    the inheritance order is critical here -- super() does not call parent, but calls the next class in the MRO
+    So, we need ConfiguredModel to take the ConfigHolder and pass kwargs to HookedTransformer
+    """
+
+    def __init__(self, cfg_holder: ConfigHolder) -> None:
+        super().__init__(
+            # for ConfiguredModel
+            zanj_model_config=cfg_holder,
+            # for HookedTransformer
+            cfg=cfg_holder.hooked_transformer_cfg,
+            tokenizer=cfg_holder.tokenizer,
+        )
+
+    def _load_state_dict_wrapper(
+        self,
+        state_dict: dict[str, Any],
+        **kwargs,
+    ) -> None:
+        """this is a wrapper around the _load_state_dict function that allows us to do extra things when loading a state dict"""
+
+        recover_exact: bool = kwargs.get("recover_exact", False)
+        fold_ln: bool = kwargs.get("fold_ln", True)
+
+        if self.zanj_model_config.model_cfg.are_layernorms_folded and fold_ln:
+            raise ValueError(
```

Comment:
Should this be an error or just a noop if we try to do it twice?

Reply:
Seems like this would trigger any time we try to load a model that was saved with folded layernorms, since `fold_ln` default to True here? And setting it to False would cause the config to be set to False which we don't want?

Reply:
Yeah, this is a good idea. I am not sure about the motivation behind folding layernorms when saving/loading, maybe @afspies knows? I think making it a noop (with warning?) would be good

Reply:
I'll just set it to `False` as default for now I think